compiler/rustc_mir_build/src/builder/scope.rs

```diff
@@ -731,43 +732,51 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
                     .unwrap_or_else(|| {
                         span_bug!(span, "no enclosing const-continuable scope found")
                     });
-                let state_place = self.scopes.const_continuable_scopes[break_index].state_place;
 
                 let rustc_middle::thir::ExprKind::Scope { value, .. } =
                     self.thir[value.unwrap()].kind
                 else {
                     panic!();
                 };
-                let rustc_middle::thir::ExprKind::Adt(value_adt) = &self.thir[value].kind else {
-                    panic!();
+
+                let scope = &self.scopes.const_continuable_scopes[break_index];
+
+                let state_ty = self.local_decls[scope.state_place.as_local().unwrap()].ty;
+                let discriminant_ty = match state_ty {
+                    ty if ty.is_enum() => ty.discriminant_ty(self.tcx),
+                    ty if ty.is_integral() => ty,
+                    _ => todo!(),
+                };
+
+                let rvalue = match state_ty {
+                    ty if ty.is_enum() => Rvalue::Discriminant(scope.state_place),
+                    ty if ty.is_integral() => Rvalue::Use(Operand::Copy(scope.state_place)),
+                    _ => todo!(),
                 };
 
-                //dbg!(&self.thir[value], value_adt);
+                let real_target = match &self.thir[value].kind {
+                    rustc_middle::thir::ExprKind::Adt(value_adt) => scope
+                        .match_arms
+                        .target_for_value(u128::from(value_adt.variant_index.as_u32())),
+                    rustc_middle::thir::ExprKind::Literal { lit, neg: _ } => match lit.node {
```

Comment:
Ignoring `neg` is not correct. That would cause `-1` to be interpreted as `1`.

Reply:
fixed that now, it is right that in MIR only unsigned values are used right, so we'd get this for `-1`

```
    bb0: {
        _1 = const 0_i32;
        _2 = copy _1;
        switchInt(copy _2) -> [4294967295: bb2, 0: bb1, otherwise: bb3];
    }
```